### ISSUE TITLE
Lesson 11 "Moving in a Circle Using Delta" not passing check if you use variables

### ISSUE THREAD

**Issue by Kazeheki** [GDQuest/learn-gdscript#541]
**Describe the bug**
If you use variables instead of doing the multiplication directly on the function calls, it doesn't pass the "Movement Is Time Dependent" check.

**To Reproduce**
```
func _process(delta):
	var rotation = 2 * delta
	var movement = 100 * delta
	rotate(rotation)
	move_local_x(movement)
```

**Expected behavior**
Same checks passing as if I entered the recommended solution
```
func _process(delta):
	rotate(2 * delta)
	move_local_x(100 * delta)
```

**Screenshots**
![grafik](https://user-images.githubusercontent.com/11597609/172609684-b9c23b66-fb6f-4b77-bf5a-f6e66a2a32cc.png)

**Information about your device:**
 - Operating System: Windows
 - Browser: Firefox
 - Build Information from bottom right on page: 2022/06/03 19:12:28 | staging::6256081abeb601d561051ad304172e216ba6dc02

**Comment by NathanLovato** [comment]
Yeah that's normal, it's a pain to check this. We don't have access to the code structure inside of the function. For now we won't try to add checks for patterns like this one.

**Comment by Kazeheki** [comment]
Would it be an option to include some hint to not use variables in that lesson?
Since a user knows about variables since lesson 7 or 8 they might want to use them.
I got confused for a moment as I didn't know why my solution was not working when it is perfectly fine code.

**Comment by NathanLovato** [comment]
Yes absolutely, we should add a code comment to mention this limitation.

**Comment by NathanLovato** [comment]
Actually, after reviewing the lessons, we can't add comments. It seems that you already have programming experience and so you wrote variables before we even introduced them. We have to assume at this stage that the student does not know what variables are, so we can't tell them not to write variables. That would just confuse them and lead to people opening issues in this repository.

**Comment by Kazeheki** [comment]
While it's true that I wrote them because I have programming experience, I think this might be a wrong assumption. In lesson 8 "Defining your own variables" students are introduced to variables.

**Comment by NathanLovato** [comment]
Ah you're right, I missed that, adding the mention now. There will still be issues because basically using local vars like that can make any practice fail in the entire course. It'll just guard a single practice. But it's better to have the mention once somewhere than never.

**Comment by droidBasher** [comment]
I have a potential fix that passes these submissions:

```
func _process(delta):
	rotate(2 * delta)
	move_local_x(100 * delta)

func _process(delta):
	move_local_x(100 * delta)
	rotate(2 * delta)

func _process(delta):
	var rotation = 2 * delta
	var movement = 100 * delta
	rotate(rotation)
	move_local_x(movement)

func _process(delta):
	var rotation : float = 2 * delta
	var move_x : float = 100 * delta * cos(rotation)
	var move_y : float = 100 * delta * sin(rotation)
	move_local_x(move_x)
	move_local_y(move_y)
	rotate(rotation)
```

but fails these submissions:

```
func _process(delta):
	rotate(2)
	move_local_x(100)

func _process(delta):
	rotate(delta)
	move_local_x(delta)

func _process(delta):
	rotate(2 *delta)
	move_local_x(0 * delta)

func _process(delta):
	rotate(2 *delta)

func _process(delta):
	rotate(100 * delta)
	move_local_x(2 * delta)
```

The approach I used was to add execution testing on _process in the existing file TestsRotatingMovingDelta.gd. I'm going to review the changes on my end until I believe it's correct, then will submit a pull request (I think that's the expected thing to do, I haven't really used github much).

**Comment by NathanLovato** [comment]
We need a general solution to the problem of using variables, and that involves writing a parser to resolve what a variable is when calling a function. Importantly, this shouldn't be solved only for one practice.

**Comment by droidBasher** [comment]
Unfortunately that sounds more difficult than what I attempted, I didn't make a general solution. Perhaps on a later weekend I can try again.

**Comment by NathanLovato** [comment]
If you´d like to try, here's one way to tackle this. You'd need to write a small tokenizer, code that can split up/abstract and store data about lines of code. It should find lines that define variables and store information about the variables in question in some data structure. Then, we should be able to spot those variable names used in other lines of code and know what their value is.

We already have the start of that in MiniGDScriptTokenizer.gd. This script can tokenize and detect infinite recursive function calls (a function calling itself) and prevents that to avoid crashing the app.

However, it's lacking support for variables at the moment. We could solve this issue using that tokenizer.

Another way would be to expose the GDScript parser from the Godot engine and write extra functions in C++ to check code, but we've decided to wait for Godot 4 before doing that as Godot 4 has a completely different parser than Godot 3, it's more reliable and easier to work with.

**Comment by Snailedlt** [comment]
+1
A built in text-to-speech narration button would also be super useful! 
For me it's much easier to digest information when hearing it as opposed to reading it.
